### PR TITLE
Replace ngram FULL_MASK Python for-loop with Triton kernel

### PR DESCRIPTION
**Pull request by shiloz-stack**
Replace ngram FULL_MASK Python for-loop with Triton kernel

## Motivation

`NGRAMWorker._prepare_for_speculative_decoding` has an explicit TODO from the original author. The `FULL_MASK` path builds the per-request `(D, seq_len + D)` tree mask with a Python `for` loop over the batch, issuing `bs` separate `torch.ones` + `torch.cat` + H2D copies per decode step. At large batch sizes this host-side loop becomes a non-trivial portion of the verify-step overhead.

This PR replaces that loop with a single Triton kernel launch.

Refs #21052.

## Modifications

- **New kernel** `python/sglang/kernels/ops/speculative/ngram_mask.py`:
  - `build_ngram_full_tree_mask` — Triton kernel. Grid `(bs, D)`, one program per draft row. Each program fills the `seq_len` history columns with `1` (tiled via `BLOCK_S`) and copies the `draft_mask[b]` block for the right-side `D` columns. Output is a flat `1 x sum(D * (seq_len_i + D))` `bool` tensor (ragged, offsets via exclusive prefix sum of row sizes).
  - `build_ngram_full_tree_mask_ref` — pure-PyTorch reference used by the unit test for bit-identical validation.
- **Worker wiring** `python/sglang/srt/speculative/ngram_worker.py`:
  - Import the new kernel.
  - Replace the `for i in range(bs)` loop with a single `build_ngram_full_tree_mask(draft_mask, seq_lens_cpu, D)` call.
- **Registry**
`python/sglang/kernels/ops/speculative/__init__.py`: registered the kernel metadata.
- **Tests** `test/registered/kernels/ops/speculative/test_ngram_full_tree_mask.py`: bit-identical test against the PyTorch reference across `bs ∈ {1, 2, 8, 32}`, `D ∈ {4, 8}`, mixed `seq_lens`, `seq_len = 0` edge case, and empty-batch guard.

## Accuracy Tests

The FULL mask is a deterministic construction (history = all-ones, right block = draft mask), so correctness is validated by **bit-identical** comparison against the PyTorch reference (`torch.equal`), not by end-to-end generation quality.
```
bs=1   D=4 seq=[0]                            ✅ PASS
bs=1   D=4 seq=[10]                           ✅ PASS
bs=1   D=8 seq=[128]                          ✅ PASS
bs=2   D=4 seq=[5, 20]                        ✅ PASS
bs=8   D=4 seq=[1, 7, 16, 64, 128, 257, 512,  ✅ PASS
bs=32  D=8 seq=[16, 32, 48, 64, 80, 96, 112,  ✅ PASS
```
## Speed Tests and Profiling

- Results from the latest run (Rev 3):

Microbenchmark on **NVIDIA A100-SXM4-40GB**, torch `2.11.0+cu128`, D=8, warmup=50, iters=200.

| bs | seq_len | elems | Python ref (ms) | Triton e2e (ms) | speedup | Triton+buf (ms) | speedup (buf) |
|----|---------|-------|-----------------|-----------------|---------|-----------------|----------------|
| 1 | 128 | 1,088 | 0.065 | 0.089 | **0.73x** | 0.084 | 0.77x |
| 8 | 256 | 16,896 | 0.308 | 0.131 | **2.35x** | 0.128 | 2.40x |
| 32 | 512 | 133,120 | 1.165 | 0.132 | **8.80x** | 0.124 | 9.37x |
| 64 | 512 | 266,240 | 2.298 | 0.131 | **17.56x** | 0.124 | 18.58x |
| 32 | 2048 | 526,336 | 1.163 | 0.131 | **8.85x** | 0.125 | 9.33x |

**Notes**
- `Python ref` = previous for-loop semantics (`build_ngram_full_tree_mask_ref`).
- `Triton e2e` = per-call `offsets` cumsum + kernel + `empty(required_numel)`.
- `Triton+buf` = same but writes into a preallocated buffer.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #30337219846](https://github.com/sgl-project/sglang/actions/runs/30337219846)
Latest PR Test (Extra): :x: [Run #30337219709](https://github.com/sgl-project/sglang/actions/runs/30337219709)

### CODE REVIEW

File: test/registered/kernels/ops/speculative/test_ngram_full_tree_mask.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import pytest
+import torch
+
+from sglang.kernels.ops.speculative.ngram_mask import (
+    build_ngram_full_tree_mask,
+    build_ngram_full_tree_mask_ref,
+)
+
+
```
Comment (zhendonghua) [inline test/registered/kernels/ops/speculative/test_ngram_full_tree_mask.py:15]: Need to register to ci like `register_cuda_ci()` and add a `__main__`

File: python/sglang/kernels/ops/speculative/ngram_mask.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    seq = seq_lens.to(device=device, dtype=torch.int32, non_blocking=True).contiguous()
+
+    # sizes[b] = D * (seq_lens[b] + D)
+    sizes = draft_token_num * (seq + draft_token_num)
+    offsets = torch.empty(bs, dtype=torch.int32, device=device)
+    offsets[0] = 0
+    if bs > 1:
+        torch.cumsum(sizes[:-1], dim=0, out=offsets[1:])
+
+    total = int(sizes.sum().item())
```
Comment (zhendonghua) [inline python/sglang/kernels/ops/speculative/ngram_mask.py]: This will enforce a D2H sync, you can use `seq_lens_cpu` to remove it

File: python/sglang/srt/speculative/ngram_worker.py [hunk 3]
```diff
@@ -328,25 +329,17 @@ def _prepare_for_speculative_decoding(self, batch: ScheduleBatch):
...
-                        req_mask,
-                        torch.from_numpy(mask[i]).to(
-                            device=self.device, non_blocking=True
-                        ),
-                    ),
-                    dim=1,
-                ).to(torch.bool)
-                tree_mask.append(req_mask.flatten())
-            tree_mask = torch.cat(tree_mask, dim=0)
+            draft_mask = (
```
Comment (zhendonghua) [inline python/sglang/srt/speculative/ngram_worker.py]: can we reuse `tree_mask`?

**Review by zhendonghua — CHANGES_REQUESTED** [review]
Please fix lint.

**Comment by shiloz-stack** [comment]
Thanks! Addressed in the latest commits:
1) `register_cuda_ci` + `__main__` on the unit test  
2) compute `total` on CPU `seq_lens_cpu` (no GPU `.item()` sync)  
3) reuse on-device `tree_mask.view(bs, D, D)`; FULL mask assigned back to `tree_mask`

Also the lint and the kernel CI got passed. Please take a look when you have chance, thanks!

File: test/registered/kernels/ops/speculative/test_ngram_full_tree_mask.py (hunk 1 above)
Comment (shiloz-stack) [reply test/registered/kernels/ops/speculative/test_ngram_full_tree_mask.py:15]: Updated

File: python/sglang/kernels/ops/speculative/ngram_mask.py (hunk 2 above)
Comment (shiloz-stack) [reply python/sglang/kernels/ops/speculative/ngram_mask.py]: Good suggestion, updated to use seq_lens_cpu for computation to avoid D2H sync

File: python/sglang/srt/speculative/ngram_worker.py (hunk 3 above)
Comment (shiloz-stack) [reply python/sglang/srt/speculative/ngram_worker.py]: Sounds good, updated to reuse `tree_mask` to avoid reassignment

File: python/sglang/srt/speculative/ngram_worker.py [hunk 4]
```diff
@@ -328,25 +329,15 @@ def _prepare_for_speculative_decoding(self, batch: ScheduleBatch):
...
-                ).to(torch.bool)
-                tree_mask.append(req_mask.flatten())
-            tree_mask = torch.cat(tree_mask, dim=0)
+            tree_mask = build_ngram_full_tree_mask(
+                tree_mask.view(
+                    bs,
+                    self.draft_token_num,
+                    self.draft_token_num,
+                ),
+                batch.seq_lens_cpu,
```
Comment (zhendonghua) [inline python/sglang/srt/speculative/ngram_worker.py]: Can we directly use `batch.seq_lens`?

File: python/sglang/kernels/ops/speculative/ngram_mask.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        draft_token_num: D.
+        block_s: Triton block size for the history fill loop.
+
+    Returns:
+        (total,) bool CUDA tensor, sum_b D * (seq_lens_cpu[b] + D).
+    """
+    if draft_mask.numel() == 0:
+        return torch.empty(0, dtype=torch.bool, device=draft_mask.device)
+
+    assert draft_mask.is_cuda, "draft_mask must be on CUDA"
```
Comment (zhendonghua) [inline python/sglang/kernels/ops/speculative/ngram_mask.py]: maybe assert not on cpu because you register your test to amd ci?

File: python/sglang/kernels/ops/speculative/ngram_mask.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    ).contiguous()
+
+    # sizes[b] = D * (seq_lens[b] + D)
+    sizes = draft_token_num * (seq + draft_token_num)
+    offsets = torch.empty(bs, dtype=torch.int32, device=device)
+    offsets[0] = 0
+    if bs > 1:
+        torch.cumsum(sizes[:-1], dim=0, out=offsets[1:])
+
+    out = torch.empty(total, dtype=torch.bool, device=device)
```
Comment (zhendonghua) [inline python/sglang/kernels/ops/speculative/ngram_mask.py]: can we pre-allocate an over-sized buffer to avoid this? cc @Qiaolin-Yu

File: python/sglang/kernels/ops/speculative/ngram_mask.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    draft_mask = draft_mask.to(dtype=torch.bool).contiguous()
+    # Compute the output size from the CPU sequence-length mirror to avoid D2H sync.
+    total = draft_token_num * (int(seq_lens_cpu.sum().item()) + bs * draft_token_num)
+    seq = seq_lens_cpu.to(
+        device=device, dtype=torch.int32, non_blocking=True
+    ).contiguous()
+
+    # sizes[b] = D * (seq_lens[b] + D)
+    sizes = draft_token_num * (seq + draft_token_num)
+    offsets = torch.empty(bs, dtype=torch.int32, device=device)
```
Comment (zhendonghua) [inline python/sglang/kernels/ops/speculative/ngram_mask.py]: can we mimic the similar way in eagle that we just calculate the cumsum outside kernel and leave other logics all to kernel?

**Review by zhendonghua — COMMENTED** [review]
Could you also update the benchmark, thanks!

File: python/sglang/kernels/ops/speculative/ngram_mask.py (hunk 5 above)
Comment (shiloz-stack) [reply python/sglang/kernels/ops/speculative/ngram_mask.py]: Updated in next rev

File: python/sglang/kernels/ops/speculative/ngram_mask.py (hunk 6 above)
Comment (shiloz-stack) [reply python/sglang/kernels/ops/speculative/ngram_mask.py]: I updated the worker to reuse `attn_backend.get_verify_buffers_to_fill_after_draft()` for the tree mask buffer to align with the pattern for eagle, so no extra allocation is needed on the hot path. Happy to switch to a dedicated pre-allocated buffer if there's a reason to keep them separate

File: python/sglang/kernels/ops/speculative/ngram_mask.py (hunk 7 above)
Comment (shiloz-stack) [reply python/sglang/kernels/ops/speculative/ngram_mask.py]: That makes sense. I cleaned up the code a bit so the cumsum computation is in the worker and passed into the kernel

**Comment by shiloz-stack** [comment]
Appreciate your feedback @zhendonghua I've addressed your comment and updated the benchmark. Please take a look when you have time and happy to discuss further
